Revamped Files plugin

Mingling94 opened the pull request:
On top of #162 sort of. That PR is awaiting changes in a sia wrapper PR: https://github.com/NebulousLabs/Nodejs-Sia/pull/14

**Sorry this PR has taken a while, but most individual commits until this point in the branch left the plugin broken or really error-prone, as would be the case with a large redesign.**
## Major changes:
- Files plugin has classes to be semi-oop!
  - Of special note are the classes `entity`, `folder`, `file` which model a virtual file system for a renter's library. `entity` isn't used except as the parent class of `folder` and `file` (and perhaps more?)
- Jqueried up Files plugin! Allows for some nifty transitions!
  - Got rid of html blueprints (hidden divs for cloning purposes) and now create elements from html-strings such as within `fileElement.js` and `folderElement.js`
- Changed how files are added to a dynamic dropdown menu, inspired by https://drive.google.com/drive/my-drive
  - Files and folders are uploaded as is, reducing the need for popups/divs asking for optional nicknames. TODO: implement easy renaming and drag&drop file/folder movement
## Minor changes:
- For this plugin only, I've messed around with styling and naming practices. They all still pass JSHint linting, but are different from how I do the rest of the UI. However, after my experiences reading other JS repos, I believe they are safe and more conventional to adopt. These changes include:
  - Only capitalize class constructors (which I don't really use since I make classes via [`Object.assign(Object.create(superclass), subclass);`](https://medium.com/javascript-scene/common-misconceptions-about-inheritance-in-javascript-d5d9bab29b0a)
  - Various es5 syntax sugars
- Actively looked for places to remove css. Things may look a bit ugly at first but css maintenance seemed slightly useful at best and arbitrary time waste at worst unless it is guaranteed to go into the next release.
## TODO's:
- File/folder drag&drop movement (as mentioned earlier)
  - Multi-select actions?
- Easy renaming
- Get search to search through all files/folders under the current folder
- Hotkeys to commonly used actions
## Screenshots:

![sia_047](https://cloud.githubusercontent.com/assets/4606838/12073753/38024784-b0fd-11e5-87f9-ae3651e4648c.png)
![sia_048](https://cloud.githubusercontent.com/assets/4606838/12073754/3d6ba788-b0fd-11e5-8c83-9a62be59c7a5.png)
![selection_050](https://cloud.githubusercontent.com/assets/4606838/12073755/3fc76ef4-b0fd-11e5-9704-82f24a5805f3.png)
![selection_051](https://cloud.githubusercontent.com/assets/4606838/12073756/464186f2-b0fd-11e5-9f93-647cb29187b2.png)
### The above shows this PR fixes #158

![sia_052](https://cloud.githubusercontent.com/assets/4606838/12073757/48d914ac-b0fd-11e5-8527-89c12216cb65.png)
![sia_053](https://cloud.githubusercontent.com/assets/4606838/12073758/4a3e3836-b0fd-11e5-8c6e-471f2c2cbb4d.png)

Mingling94 committed:
Jqueried up Files plugin

Also removed code from index.html

Mingling94 committed:
Drafted up File object for file library

Mingling94 committed:
Drafted up folder and entity modules for browser

Mingling94 committed:
Touched up entity, file, folder objects

Mingling94 committed:
Add sia icon by default to message dialogs

Mingling94 committed:
Fixed improper Hosting rebase about Siad.configure

Mingling94 committed:
Allow es5 syntax

Mingling94 committed:
Put button clickability in the global js file

Mingling94 committed:
Lots of changes through browser code

Have yet to test if this works.

Further jqueried up lots of DOM code.

Implemented browser a bunch more. Need to add buttons to use browser
navigation

Added uiTools library to offer shortcuts to ipc functionalities like
notify, tooltip, and dialog

Mingling94 committed:
Slight config changes

Mingling94 committed:
Revamped Upload/Add buttons

Mingling94 committed:
Separated out updateBrowser and updateList

Added folderElement.js
Added size getter to file, folder, and entity
Reorganized other code

Mingling94 committed:
Combined browser.js and loaders.js

Enables the button-options from the new file/folder button

Mingling94 committed:
Small fixes

Mingling94 committed:
Various fixes, files show up now

Mingling94 committed:
Folders now work

Mingling94 committed:
Current path shows at header, various fixes

Fixed various errors throughout the Files plugin and drafted up the
current working directory at the top

Mingling94 committed:
Delegated folder creation to folders themselves

Mingling94 committed:
Made folders clickable to navigate to that folder

Mingling94 committed:
Got rid of font-size specifiers

They're troublesome

Mingling94 committed:
Renamed cb to callback in folder.js

Mingling94 committed:
Multiselect uploading files, folders, and .sia's

Mingling94 committed:
Homogenized Files & Hosting's top-of-file comments

Mingling94 commented:
@DavidVorick @lukechampine @VoidingWarranties 
I invested quite a bit of time into this and I'd appreciate a thorough code review. Especially consider that I'm experimenting with some new JS conventions/tools/architecture things that I'm totally open to input/criticism on.

I understand there's 2k+ line changes and that might take a while. I'll be developing on top of this PR in another branch.

DavidVorick commented:
Great to see this, thanks. I will try to look at it asap, though I can't promise too much in terms of quality because I'm not very good with javascript. I will give it my best though

lukechampine commented:
Maybe it's just unclear from the screenshots, but can you nest folders within folders? If so, we'll need some way of displaying the user's current path within the directory structure. Google Drive again seems like a good place to draw inspiration from.

Mingling94 commented:
@lukechampine The part where it says `New Folder` is meant to be a CWD just like with GDrive and could be `New Folder/New Folder`. I just haven't gotten it working yet. That's next up and reserved for another PR on top of this one.

DavidVorick commented on .jshintrc:
```diff
@@ -33,7 +33,6 @@
 	"boss"          : false,    // Tolerate assignments inside if, for & while. Usually conditions & loops are for comparison, not assignments.
 	"debug"         : false,    // Allow debugger statements e.g. browser breakpoints.
 	"eqnull"        : false,    // Tolerate use of `== null`.
-	"es5"           : false,    // Allow EcmaScript 5 syntax.
```
in the future, better to have a change like this be a separate PR.

Mingling94 replied:
ES5 standards are used heavily throughout the PR

DavidVorick commented on js/mainjs/addIPCListeners.js:
```diff
@@ -30,7 +34,10 @@ module.exports = function(config, mainWindow) {
 				console.error('Unknown dialog ipc');
 		}
-		event.returnValue = response ? response : null;
+		// TODO: Make async and adapt plugins
```
can you explain this todo a bit better? I'm guessing there's not going to be enough time to do this.

Mingling94 replied:
It's not that hard. Currently dialog windows block the UI, a minor annoyance but makes for slightly easier to read code. From the [dialog module in the electron documentation](http://electron.atom.io/docs/v0.36.0/api/dialog/):

> If a callback is passed, the API call will be asynchronous and the result will be passed via callback(filenames)

Mingling94 committed:
Reformatted code and centralized waterfall code

Mingling94 committed:
Fixed some bugs

folder.parentFolders was improperly coded

Exposed currentFolder and rootFolder for debugging

Clicking folder buttons navigated to them

CSS for file path is more usable

Mingling94 committed:
Recursive upload, untested so far

Mingling94 committed:
entityElement.js

fileElement.js and folderElement.js have a lot of common code. I put
that in entityElement.js

Mingling94 committed:
Added label entry

Mingling94 committed:
Introduce selected class/member

Mingling94 committed:
Made file selection system

Gave file, entity, and folder actual getters

Mingling94 committed:
File selection system shift and ctrl click actions

Mingling94 committed:
Adapted Files to new api

Mingling94 committed:
More valuable error logging from api calls

Mingling94 committed:
Separated out loading functions from browser.js

Mingling94 committed:
Applied waterfall function to folder uploading

Mingling94 committed:
Got rid of entity.js, folder inherits from file

Mingling94 committed:
Fixed dialog. It sometime blocked renderer process

with no returnValue, a synchronous ipc call never returned

Mingling94 committed:
New button is now just last directory

pmknutsen commented:
Nice job on the plugins! I look forward to testing this in a few days. I have some comments on the screenshots:

"New" context-menu:
Display menu below and left-aligned with New. it should not cover New. I think the plus is redundant.

For consistency, I suggest to style popup menus as popup notifications: straight, gray borders w/shadows. Horizontal dividers in popups should be a fainter gray than text.

Fix popup notification clutter:
When selecting multiple files/folders for upload, display a _single_ popup saying: "Added N files".

Feature request:
Deleting multiple files is a hassle (click one, confirm one, repeat, repeat, repeat...). Options:
1) Click on a filename, highlight it and remove with Delete key.
2) Instead of delete icons, display a checkbox after each file and a Delete option in the plugin toolbar.

Files / folders list:
1) Display folders before files.
2) Sort files alphabetically (later, add option to sort by filesize, time remaining like a conventional file browser).  

When user is not in the root folder, display a back arrow rather than folder icon in toolbar.

Throughout, I think text in the plugin frames need to be a bit darker. Text appears faint on some monitors.

Mingling94 committed:
Lots of changes

* Selection system has 'anchor'
* Folder inherits from File, Entity is gone
* Can rename stuff by clicking the name
* New file/folder button is gone, dropdown menu appears from last
clicking folder

Mingling94 committed:
Made anchor system more concise

Mingling94 commented:
Aight, the PR is at a fairly stable point now, though I'm not trying to put forth many more hours to it today. I referenced many core things still yet TODO: in this issue: https://github.com/NebulousLabs/Sia-UI/issues/174
